Paper: High-performance operator evaluations with ease of use: libCEED's Python interface

valeriabarra opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

valeriabarra committed:
Initial commit for libCEED's SciPy2020 paper submission - Add README

valeriabarra committed:
Update README.rst

valeriabarra committed:
Add text from abstract submission

valeriabarra committed:
Have only paper.rst in top-level paper directory as per their instructions

valeriabarra committed:
Use format from conference template

Note: figure labels do not work as shown in their examples https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst

They were removed, otherwise it would not compile

valeriabarra committed:
Mention petsc4py as an example of external package for parallel decomposition

valeriabarra committed:
Add Orszag citation and some edits to abstract and Intro

valeriabarra committed:
Better describe decomposition and action of operator

valeriabarra committed:
Better explain GPU support in our currently available C QFunctions

valeriabarra committed:
Mention both Numba and Jax to possibly use for Python QFunctions

valeriabarra committed:
Another attempt at describing the decomposition and action of operator

valeriabarra committed:
Python -> Python3 in make_paper.sh

This is needed on Ubuntu, hopefully it won't screw anything on your systems, if that is your default Python

valeriabarra committed:
Introduce Numba CUDA array interface for device memory handling

valeriabarra committed:
Expand intro and add sketch of libCEED's role with different backends and user codes

valeriabarra committed:
Add figure labels and reference them as Fig. in the text

valeriabarra committed:
Add descriptions of T-vec, L-vec, E-vec and Q-vec and code block examepls

valeriabarra committed:
Add references for website citations

valeriabarra committed:
Reword: interface -> API

valeriabarra committed:
Add Conclusions and Acknowledgments

valeriabarra committed:
Add CEED MS citations

valeriabarra committed:
Add libCEED dev site citation

jedbrown committed:
Trim figure so it doesn't overlap adjacent column

jedbrown committed:
Revise abstract

jedbrown committed:
Spell-check

jedbrown committed:
Copy editing/todos

valeriabarra committed:
Add Lottes-Fischer reference

valeriabarra committed:
Final edits

jedbrown committed:
Editing...

jedbrown committed:
Editing

valeriabarra committed:
Remove two CEED Technical Report references

valeriabarra committed:
Some edits suggested by @jedbrown

Moved CEED software distribution citations to last sentence before Conclusions

Mentioned JIT together with device Numba/JAX implementations

Rephrased beginning of libCEED's API section

Also changed one instance of flops -> FLOPs

valeriabarra committed:
Add :corresponding: field

valeriabarra committed:
Merge pull request #1 from valeriabarra/libceed-scipy-paper

Libceed scipy paper

deniederhut commented on make_paper.sh:
```diff
@@ -7,7 +7,7 @@ if [[ ! -d $DIR ]]; then
 fi
 
-python publisher/build_paper.py $DIR
+python3 publisher/build_paper.py $DIR
```
We prefer that paper PRs not modify any files outside of their own directory. Would you mind reverting this change?

valeriabarra replied:
Absolutely, sorry. I meant to do that already but it must have slipped.

deniederhut replied:
No worries - thanks for making the change!

valeriabarra committed:
Revert python3 -> python in make_paper.sh

vahtras commented:
Hi, this is to introduce myself as a reviewer. Not done yet, but looks very good so far

valeriabarra commented:
Thanks for the introduction @vahtras  and thanks for your review!

vahtras commented:
Regarding the code snippets, we now have a mixture of bash and Python with the same font/notation. This is unfortunate. I strongly recommend the established practice of writing out the prompts in both cases, e.g.

    $ pip install libceed

vs

    >>> import libceed

i.e. all Python examples written as interactive sessions. If an output is expected from a Python instruction (e.g. print) it should be written out, such that it can be verified with doctest

valeriabarra commented:
Sure, thanks for pointing that out. I will take care of it.

valeriabarra committed:
libCEED paper review feedback: write prompts in code snippets

valeriabarra commented:
Thanks for the feedback. 
For now I have added the prompts in the snippets only. I will add the outputs and try to verify with doctest

deniederhut commented:
Hey @valeriabarra ! Thanks to some awesome behind the scenes work by @stargaser, we now have experimental support for publishing papers with ORCIDs. If you have one and would like to add it to your paper, it's as simple as adding an `:orcid:` tag under your name in the paper header. [You can see an example of how to do this in the `00_vanderwalt` example paper](https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2020/papers/00_vanderwalt/00_vanderwalt.rst). If you don't have one, no worries! You can still publish your paper as is.

valeriabarra committed:
Add ORCIDs

valeriabarra commented:
Thanks @deniederhut! I added them

valeriabarra committed:
Add code snippet outputs

Once the wheels for the pip distribution are updated, this should be doctest clean with the following test line

doctest.testfile("papers/libceed-paper/paper.rst", optionflags=doctest.NORMALIZE_WHITESPACE)

The optionflags=doctest.NORMALIZE_WHITESPACE is needed because the whitespaces in the output of the basis print statement did not match the expected behavior, because of a combination of whitespaces and tabs

deniederhut commented:
Hi @vahtras ! Did you have further feedback on this paper?

valeriabarra committed:
Doctest: remove <BLANKLINE>. This can be avoided with optionflags=doctest.NORMALIZE_WHITESPACE

vahtras commented:
I am pleased with the result and I have not more comments

deniederhut commented on papers/libceed-paper/bibliography.bib:
```diff
@@ -0,0 +1,198 @@
+    year = "2020"
+}
+
+@misc{NumPy,
```
Could I ask you to update the citation for NumPy? https://ieeexplore.ieee.org/document/5725236

deniederhut commented on papers/libceed-paper/bibliography.bib:
```diff
@@ -0,0 +1,198 @@
+  number = {ANL-95/11 - Revision 3.13}
+}
+
+@misc{ProjectJupyter,
```
Could we ask you to update the citation for Jupyter? http://ebooks.iospress.nl/publication/42900

deniederhut commented on papers/libceed-paper/bibliography.bib:
```diff
@@ -0,0 +1,198 @@
+    year = "2020"
+}
+
+@misc{Numba,
```
Could I ask you to update the citation for numba? https://dl.acm.org/doi/10.1145/2833157.2833162

valeriabarra replied:
Sure thing. I updated them. Thanks for pointing out the appropriate references.

valeriabarra committed:
Update citations

deniederhut commented:
Sweet! I think this is good to go.

cbcunc reviewed (approved):
Passes final proceeding committee walk-through.
